common/tasks/multi_destination_scheduler.go

```diff
@@ -0,0 +1,241 @@
+		lim.reservations++
+		return true
+	}
+	return false
+}
+
+// release a reservation.
+func (lim *limiter[T]) release() {
+	lim.reservations--
+}
+
+// MultiDestinationSchedulerOptions are options for creating a [MultiDestinationScheduler].
+type MultiDestinationSchedulerOptions[K comparable, T Task] struct {
+	// A function to determine the destination of a task.
+	KeyFn func(T) K
+	// Limiter options for a given destination.
+	LimiterOptionsFn func(K) LimiterOptions
+	Logger           log.Logger
+}
+
+// MultDestinationScheduler for limiting concurrency and RPS (and later applying a circuit breaker) on a per destination
+// basis. A task's destination is determined based on a provided key function.
+// Accepts and buffers tasks that exceed the destination's concurrency limit to be picked up as soon as there's more
+// availability.
+type MultiDestinationScheduler[K comparable, T Task] struct {
+	options      MultiDestinationSchedulerOptions[K, T]
+	waitGroup    sync.WaitGroup
+	stopCtx      context.Context
+	stopFn       context.CancelFunc
+	limitersLock sync.RWMutex
+	limiters     map[K]*limiter[T]
+}
+
+func NewMultiDestinationScheduler[K comparable, T Task](options MultiDestinationSchedulerOptions[K, T]) *MultiDestinationScheduler[K, T] {
+	stopCtx, stopFn := context.WithCancel(context.Background())
+	return &MultiDestinationScheduler[K, T]{
+		options:  options,
+		stopCtx:  stopCtx,
+		stopFn:   stopFn,
+		limiters: make(map[K]*limiter[T]),
+	}
+}
+
+func (*MultiDestinationScheduler[K, T]) Start() {
+	// noop
+}
+
+// Stop signals running tasks to stop, aborts any pending tasks and waits up to a minute for all running tasks to
+// complete.
+func (s *MultiDestinationScheduler[K, T]) Stop() {
+	s.stopFn()
+	s.abortBufferedTasks()
+	if success := common.AwaitWaitGroup(&s.waitGroup, time.Minute); !success {
+		s.options.Logger.Warn("MultiDestinationScheduler timed out waiting for tasks to complete")
+	} else {
+		s.options.Logger.Debug("MultiDestinationScheduler drained")
+	}
+}
+
+func (s *MultiDestinationScheduler[K, T]) abortBufferedTasks() {
+	s.limitersLock.Lock()
+	defer s.limitersLock.Unlock()
+
+	for _, lim := range s.limiters {
+		lim.Lock()
+		for _, task := range lim.buffer {
+			task.Abort()
+		}
+		lim.buffer = nil
+		lim.Unlock()
+	}
+}
+
+// Submit is not implemented as it is never invoked.
+func (*MultiDestinationScheduler[K, T]) Submit(T) {
+	panic("unimplemented")
+}
+
+// TrySubmit submits a task for processing unless the scheduler is shut down.
+func (s *MultiDestinationScheduler[K, T]) TrySubmit(task T) bool {
+	// Add to the wait group before checking if we're stopped to prevent Stop() from returning early if TrySubmit is
+	// called concurrently.
+	s.waitGroup.Add(1)
+	if s.stopCtx.Err() != nil {
+		s.waitGroup.Done()
+		return false
+	}
+
+	if lim := s.limiterForTask(task); lim != nil {
+		go func() {
+			defer s.waitGroup.Done()
+			s.executeUntilReleased(task, lim)
```

Comment:
This looks like a common go mistake. You should pass lim as an argument to the func you start with go. You should also check your tests for MultiDestinationScheduler, since the tests did not catch this error.